Python: Autogenerate Python functions corresponding to the ones in SQL

binste opened the pull request:
This PR adds the script `generate_functions_code.py` which generates `duckdb/func.py` based on the json files in `src/core_functions`. As discussed in https://github.com/duckdb/duckdb/discussions/12134, these Python functions provide a convenient API for users to create `FunctionExpression`s, similar to what you would have in Polars, Ibis, etc. CC @Tishj: I've opened the PR after all already so you can comment if you have the time and I could work on your feedback in 1-2 weeks.

# Example
```python
import duckdb
from duckdb import func
from duckdb import ColumnExpression as col
from duckdb import ConstantExpression as lit
import pandas

input_df = pandas.DataFrame.from_dict(
    {"i": [1, 2, 3, 4], "j": ["one", "two", "three", "four"], "k": [4, 3, 2, 1]}
)

rel = duckdb.from_df(input_df)

# Would be great if there is a way to add a column instead of
# having to mention all again?
rel.project(
    func.today().alias("today"),
    func.string_split(col("j"), lit("r")).alias("j_split"),
    "i",
    "k",
    func.greatest(col("i"), col("k")).alias("greatest_i_k"),
)
```

Screenshots of the IDE support this gives:
<img width="689" alt="image" src="https://github.com/duckdb/duckdb/assets/23366411/b84391fb-55f0-4233-9918-841a2c0696e7">

<img width="567" alt="image" src="https://github.com/duckdb/duckdb/assets/23366411/8d1823cb-d322-4cf8-aac8-695e0aabe774">


# Open questions
Open questions I'm aware of:
- [ ] I've excluded functions which require a lambda as I didn't find a way to express this with a `FunctionExpression`. Do you know of a way to already use these functions with the Expression API or would it require an extension? Is there a way to pass a lambda function as a string?
- [ ] Also excluded `table`, `macro` and `pragma` functions as I don't think they can be used anywhere in the Python API right now. Correct?
- [ ] Is `func` a good module name? It's what [sqlalchemy uses](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.func). I think it works well but that's just a personal preference. It could also be `f`, `fct`, ...
- [ ] Did I add the `generate...` script in the right place in the build process? I did not test this as I'm not sure how it works.

binste committed:
First version of generate_functions_code

binste committed:
Add to makefile

binste committed:
Make type hints compatible with Python 3.7 (lowest supported version by duckdb)

binste committed:
Add disclaimer in beginning of file

binste committed:
Add tests

binste committed:
Add test_aliases

binste committed:
Export 'func'

binste committed:
Format code

binste committed:
Use Optional instead of '|' for compatibility with earlier Python versions

binste committed:
Replace str.removesuffix with something that works on earlier versions of Python as well

binste committed:
Merge branch 'main' into autogenerate_python_functions

binste committed:
Format code

binste committed:
Add two new lines between function defs

binste committed:
Add new line to end of file and remove hanging commas if no parameters

binste committed:
Format code

binste commented:
I'd need help getting the CI to pass. The format script seems to be happy but it does not detect the file in the `scripts` folder:

![image](https://github.com/duckdb/duckdb/assets/23366411/e5a0e5a8-2777-4509-b205-6e66f8886b4e)

If I format that separately, everything looks good:

![image](https://github.com/duckdb/duckdb/assets/23366411/99f96a24-7b94-4eb8-86d3-b06864578e64)

Tishj commented:
> I'd need help getting the CI to pass. The format script seems to be happy but it does not detect the file in the `scripts` folder:
> If I format that separately, everything looks good:

We generally use `make format` (or `make format-main`), that should also pick up files in the `scripts` folder I would imagine?

Tishj commented:
Some functions might indeed be missing, some more context behind the `functions.json` files:
<https://github.com/duckdb/duckdb/pull/7310>

It does not include all functions
It might be better to instead use `duckdb_functions()` to feed the script

As that is guaranteed to contain everything

binste commented:
> Details

Hmm yeah I'm running `make format-main` and `make format` locally and it looks good. But my statement was wrong. It detects files in `scripts`. Turns out the error is in the `Generated Check` step of the workflow. If I run `make generate-files` locally, there are no uncommitted changes: https://github.com/duckdb/duckdb/actions/runs/9412349182/job/25927094569?pr=12270 However, in the pipeline it shows a diff for the whole `duckdb/func.py` file?

binste committed:
Merge branch 'main' into autogenerate_python_functions

binste committed:
Rewrite to use duckdb_metadata function

Tishj commented:
> > Details
> 
> Hmm yeah I'm running `make format-main` and `make format` locally and it looks good. But my statement was wrong. It detects files in `scripts`. Turns out the error is in the `Generated Check` step of the workflow. If I run `make generate-files` locally, there are no uncommitted changes: https://github.com/duckdb/duckdb/actions/runs/9412349182/job/25927094569?pr=12270 However, in the pipeline it shows a diff for the whole `duckdb/func.py` file?

I imagine it's a version difference
Mine are:
```
black==24.3.0
cmake-format==0.6.13
clang-format==11.0.1
```

binste committed:
Fix various issues

binste committed:
Remove notebook

binste committed:
Fix order of parameters. Remove internal functions

binste committed:
Replace single quotes with double quotes

binste committed:
Also test if parameters which are not passed to FunctionExpression are None

binste committed:
Update docstring of test

binste committed:
Exclude duckdb/func.py from being formatted as it's an autogenerated file

binste committed:
Increase number of expected functions

binste committed:
Use consistent parameter names across combinations when deduplicating them

binste committed:
Clarify code

binste committed:
Simplify code and improve error handling (now it relies on duckdb to surface error messages) by using UNDEFINED singleton

binste committed:
Add a simple query as a test

binste committed:
Format code

binste commented:
Using `duckdb_functions` is indeed much better! A lot of functions were missing in the json files and it now seems that we also have all combinations of parameters (i.e. function signatures).

I've reworked the code generation and it's ready to be reviewed. See some open questions in the initial post which I updated.  Also:
* You can find some example code in a test I've added
* I’ve extended the `file_is_generated` function in `format.py` so that `duckdb/func.py` is not formatted, in line with other autogenerated files.

binste committed:
Format code

binste committed:
Change list to List in type hints

binste committed:
Fix incompatible type hint '|'

binste commented:
Just a friendly reminder that I'm happy to work on this in case there is interest and someone can help me out with these questions:

> Workflows fail due to these 2 errors as far as I see it. I think it'd be much more efficient to get your input on this as it would take me some time to dig through it. Hope that's ok:
> 
> * `CodeQuality / Format Check` fails as it can't call the duckdb cli. How can I do this in the CI/CD pipeline? See `check_output` in the `get_duckdb_functions` function in `generate_functions_code.py`.
> * `Python 3.9 Linux` failed in a previous iteration as there are no stubs for `duckdb/func.py`. I'd probably just need to rerun the stub generation process but it says that there are also some manual adjustments needed.

Tishj commented on Makefile:
```diff
@@ -435,6 +435,7 @@ generate-files:
 	python3 scripts/generate_serialization.py
 	python3 scripts/generate_enum_util.py
 	python3 tools/pythonpkg/scripts/generate_connection_code.py
+	python3 tools/pythonpkg/scripts/generate_functions_code.py
```
As you mentioned, this fails because it relies on the `duckdb` module being present.
Either we need to change the CI run that uses `make generate-files` to first build python duckdb, I think this makes sense if there is only one workflow that does this.

binste replied:
Yeah agree, efficient if the build only happens in one place.

[Here](https://github.com/duckdb/duckdb/discussions/12134#discussioncomment-9556222) you mentioned, that the plan is to move `pythonpkg` into a separate repo which would probably help for this. In the meantime, a pragmatic solution could be that I:
* deactivate the execution of `generate_functions_code.py` in the `CodeQuality` workflow
* add it, incl. the `git diff` check, in the `Python 3.9 Linux` workflow after the package is built

This way, we'd at least still have something that checks that `func.py` is kept in sync. Thoughts?

Tishj commented:
For the stubs, that comment you read is kind of outdated, stubs aren't generated, they have to be manually edited.
Build the package and run `python3 -m pytest tools/pythonpkg/tests/stubs` to see what it reports, if it mentions methods you edited/added you need to update the stubs

binste committed:
Merge branch 'main' into autogenerate_python_functions

binste committed:
Add func to stub

binste committed:
Replace 'FunctionExpression' return type with actual type 'Expression'

binste committed:
Fix test_func

binste committed:
Move 'func' to top of __init__ to harmonize it

binste commented:
> For the stubs, that comment you read is kind of outdated, stubs aren't generated, they have to be manually edited. Build the package and run `python3 -m pytest tools/pythonpkg/tests/stubs` to see what it reports, if it mentions methods you edited/added you need to update the stubs

What I did:
1. I've followed the installation instructions with `pip` in `tools/pythonpkg/README.md`
2. If I then run `test_subs` locally, it tells me that `duckdb-stubs.typing` shadows the built-in `typing` module of the standard library.
3. If I move everything in `duckdb-stubs` into a subfolder `duckdb-stubs/duckdb` it works and passes. I did not push these changes as in CI there is even another error:

```bash
tmp/tmp.UyYfmKuO7b/venv/lib/python3.9/site-packages/duckdb-stubs/__init__.pyi:9: error: Skipping analyzing "duckdb.func": module is installed, but missing library stubs or py.typed marker  [import-untyped]
```
<img width="1071" alt="image" src="https://github.com/user-attachments/assets/8d4c4a60-73b4-4353-a1a2-bcd7c7ea2560">

I don't think `duckdb.func` needs a separate stubfile, it already has type hints in the file itself. Same as `duckdb/filesystem.py` which also does not have a `.pyi` file but somehow it does not throw an error for this.

Regarding the `duckdb-stubs/duckdb` subfolder, this is also generated if I run `scripts/regenerate_python_stubs.sh`.

Sorry for constantly switching this PR from "Draft" to "Ready for review" but it's tricky to debug as I can't reproduce it locally. Any ideas what's going on when looking at the error above?

Tishj commented:
Sorry for only looking at this now, this PR sort of slipped my mind
I can vouch from experience that stubs are indeed hell, I'll check out the branch and see what needs to be done to get this over the finish line

Tishj committed:
Apply suggestions from review. Allow script to fail in make generate-files step, remove duplciated functions, avoid circular imports

binste commented:
Thank you @Tishj! I've applied your diffs. Unfortunately, there is still an issue with the stubs :(
